### ISSUE TITLE
BC Break: Module Globals

### ISSUE THREAD

**Issue by kooldev** [trowski/fiber-rfc#9]
The RFC does not address the impact of fibers on the use of module globals. Extensions that use module globals for tracing (XDebug for example) have to be rewritten in a fiber-aware way or they will segfault when more than a single fiber is used. This requires you to introduce some kind of fiber-local storage in addition to the existing module globals. The alternative to that is to have each extension deal with it individually (maybe by using HashMaps with the current fiber ID / address as key) but it still needs to be mentioned in the RFC.

In addition to the major BC break this will also have a negative impact on performance because access to fiber-local data will be (at least a bit) slower than what we have today.

**Comment by kelunik** [comment]
I think this is also a good point why it should be part of `php-src`.

**Comment by kooldev** [comment]
@kelunik Aggreed. It is definitely a more convincing reason for including it into PHP than the reasons that are given in the RFC. While fibers may work (depending on used extensions / functions) as an extension they need core support to be reliable. Errors caused by unexpectd access to module from different fibers can segfault (if you are lucky) but they can also go undetected and access corrupted data without a warning.

It would probably be a good idea to create an API similiar to module globals for that purpose. Add a set of macros to declare the needed struct, compute their size and register them with fibers. Whenever a new fiber context is created allocate some additional memory (fiber struct size + sum of all fiber-local struct sizes) to avoid additonal memory allocations. To access them use a (pre-computed) offset into the fiber context. Basically a fiber-based version of what `TSRM.h` provides in PHP (it might also need constructors / destructors for fiber-local data).

**Comment by Danack** [comment]
> It would probably be a good idea to create an API similiar to module globals for that purpose.

Yeah...the less work there is for maintainers, the less whining there would be.

I would imagine that a reasonably clever implementation would have really quite a low overhead.

> why it should be part of php-src

Just to be a pedant, it's why it needs to be part of php-src, rather than should be. Small difference, but fewer places where people can pretend to be fauxrage'd by an RFC the better.

**Comment by trowski** [comment]
@kooldev I haven't observed the segfault behavior you describe with Xdebug and ext-fiber. Could you explain further what the issue is and how it could be fixed so I can add it to the RFC? Perhaps a way to replicate it?

**Comment by kooldev** [comment]
@trowski The bug seems to have been fixed in xdebug/xdebug#701. I can confirm that it works with latest XDebug `master` now and the generated code coverage html report is looking fine.

I did some more testing and recorded a profile to illustrate how code that is not aware of fibers perceives PHP code execution:

<img src="https://gist.githubusercontent.com/kooldev/3e27252dd4cd66394c1df5251f25d7a6/raw/48331266be5e03b39ebf5cef0ea53f0c50af104b/Profile.png" alt="" />

```php
$scheduler = new class() implements FiberScheduler {

    private $ready = [];

    public function enqueue(Fiber $fiber): void
    {
        $this->ready[] = $fiber;
    }

    public function run(): void
    {
        $count = 0;

        while ($this->ready) {
            $fiber = array_shift($this->ready);

            if ($fiber->isSuspended()) {
                if (++$count != 4) {
                    $fiber->resume();
                } else {
                    $fiber->throw(new \RuntimeException('Fail'));
                }
            } else {
                $fiber->start(++$count);
            }
        }
    }
};

$q = function (Fiber $fiber) use ($scheduler) {
    $scheduler->enqueue($fiber);
};

$job = function (int $i) use ($q, $scheduler) {
    echo "JOB {$i}: Start\n";
    try {
        Fiber::suspend($q, $scheduler);
        echo "JOB {$i}: OK\n";
    } catch (\RuntimeException $e) {
        echo "JOB {$i}: {$e}\n";
    }
};

$scheduler->enqueue(Fiber::create($job));
$scheduler->enqueue(Fiber::create($job));

var_dump('SUSPEND');
Fiber::suspend($q, $scheduler);
Fiber::suspend($q, $scheduler);
var_dump('RESUME');
```

Both "outline" and "Stack trace" show that maintaining a single stack across fiber boundaries can lead to false results. Profilers usually employ such a stack to record execution time of function calls. It also affects PHP features like output buffering and registration of userland error handlers.

It is up to you if you want to provide a solution to that problem along with fibers or let everbody deal with it in their own way. The RFC has to provide some insights into this and clearify what the expected behavior is (for builtin features) and what extension authors need to know to be compatible with fibers.

**Comment by kelunik** [comment]
Profilers are a special case that needs attention.

Output buffering is mostly unaffected, but unfortunately it employs a stack like the error handlers. They probably should be made fiber locals. Good point!

**Comment by trowski** [comment]
I'm not sure output buffering needs to be fiber local. I think either approach way might produce unexpected results, but having different output buffers for each fiber adds a lot of complexity. Amp has essentially used green-threads without the need for an independent output buffer in each green-thread.

**Comment by kelunik** [comment]
Yeah, because we don't use output buffering at all. ;-)

**Comment by trowski** [comment]
Userland could make a fiber-aware output buffering system if it was necessary, so I think I'll leave this out of internals.

**Comment by kelunik** [comment]
@trowski Error handling needs to be solved at least, because the usual `try { ... } finally { restore_error_handler(); }` won't work reliably if the fiber is suspended in the block.

**Comment by trowski** [comment]
This raises a question for me: At what point is it the responsibility of the developer vs. the responsibility of the engine? I think process globals such as the error handler should not be fiber-specific, otherwise we go down the path of everything needing to be fiber-specific and it starts to feel more like threading.

**Comment by kelunik** [comment]
It's fine with globals, but I think the few exceptions where the engine provides a stack, it's the responsibility of the engine, because there's no way a user can reasonably implement it.
